### PR TITLE
[WIP] Danger bot outputs test failures in detail in PR

### PR DESCRIPTION
**Pull request by ananyo2012**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [ ] all tests pass -- `rake test:all`
* [X] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by ananyo2012** [comment]
@jywarren I have set the test reporters for rails tests. Also added a failing test to show how danger reports it. I will remove the failing test once yo have approved it. Any more changes requested?

**Comment by jywarren** [comment]
Oh, interesting!! Do you think it's possible to link to the error itself, or to reproduce the error output in the Danger response? This looks really great!

**Comment by ananyo2012** [comment]
@jywarren Well the danger plugin reads the test report generated in xml format. I used [ci_reporter](https://github.com/ci-reporter/ci_reporter) for generating the test reports. Changing the test report format in ci_reporter may do the trick. A description field needs to be added.

**Comment by jywarren** [comment]
Maybe attaching a Gist of the XML output would help us figure out how to parse the exact line to link to, or the error output?

**Comment by jywarren** [comment]
Hi, @ananyo2012 - just checking in on if you're stuck on this. Do you have an easy way to post the XML output so we could try to figure out how to mine it?

**Comment by ananyo2012** [comment]
Hi @jywarren will work on this today. Didn't had time to check this out.

**Comment by ananyo2012** [comment]
Here is the xml report generated for AdminController test by  [ci_reporter](https://github.com/ci-reporter/ci_reporter) https://gist.github.com/ananyo2012/5f74fadf8a15b76dfc669c1e39337d44.

We need to add the test description field in the report.

**Comment by jywarren** [comment]
Resolved conflicts and attempted an outputting of `junit` failures via docs at https://github.com/orta/danger-junit#junit -- let's see how this goes!

File: .travis.yml [hunk 1]
```diff
@@ -15,8 +15,10 @@ install:
...
   - docker-compose exec web bash -c "rake db:migrate"
   - docker-compose exec web bower install --allow-root
-  - 'if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then gem install danger; fi'
+  - 'if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then gem install danger danger-junit; fi'
 
 script:
-  - docker-compose run -e CI=TRUE web bash -c "rake test:all"
+  - docker-compose run -e CI=TRUE web bash -c "GENERATE_REPORT=true rake test:all"
+  - echo -e '<?xml version="1.0" encoding="UTF-8"?>' > output.xml
+  - tail -n +2 -q ./test/reports/TEST*.xml >> output.xml
```
Comment (jywarren) [inline .travis.yml]: Maybe we need to leave a comment to explain this line.

**Comment by jywarren** [comment]
Cool! I want to try to parse out the path to the line of test code that failed and make a link to it in the code of the PR branch. Should take some regex trickery but be possible.

File: .travis.yml (hunk 1 above)
Comment (ananyo2012) [inline .travis.yml]: @jywarren What this does is concatenates all the xml reports into one file that is used by junit to parse the result. ci_reporter generates xml reports for each test file.

File: Dangerfile [hunk 2]
```diff
@@ -26,3 +26,9 @@ if github.pr_body.include?("* [ ]") && !github.pr_title.include?("[WIP]")
 end
 
 message "Pull Request is marked as Work in Progress" if github.pr_title.include? "[WIP]"
+
+junit.parse "output.xml"
+junit.report
+junit.failures.collect(&:nodes).flatten.each do |failure|
+  message "Failure: #{failure.nodes.inspect}"
```
Comment (ananyo2012) [inline Dangerfile]: Can you make this warn ?

**Comment by jywarren** [comment]
I made it `error` -- but the path won't be exactly right -- we need the branch the PR is being requested from, not `http://github.com/publiclab/plots2/tree/master` -- but let's see how this does, and we can make that change next.

File: test/functional/admin_controller_test.rb [hunk 3]
```diff
@@ -388,4 +388,8 @@ def teardown
     assert_equal 1, node.author.status
     assert_redirected_to node.path(:question)
   end
+
+  test 'fail a test' do
```
Comment (ananyo2012) [inline test/functional/admin_controller_test.rb]: Just to keep this as a reminder we should remove this test once we have completed this

**Comment by jywarren** [comment]
Using http://rubular.com/r/lDEp0Lx8Q1

File: .travis.yml (hunk 1 above)
Comment (jywarren) [inline .travis.yml]: oh rad. adding as a comment.

**Comment by ananyo2012** [comment]
@jywarren Something like `http://github.com/{github.pr_author}/plots2/tree/{github.branch_for_base}` should work

**Comment by ananyo2012** [comment]
Wonder why this isnt working

**Comment by jywarren** [comment]
I read in the cited danger.rb (in the danger gem source) and looks like `path` is a reserved word -- go figure! Changed to `source_path` -- let's see, fingers X!

**Comment by ananyo2012** [comment]
Great! The regexp is working. The link is broken though. Extrapolation syntax is wrong. It should be `http://github.com/#{github.pr_author}/plots2/tree/#{github.branch_for_base}`

File: Dangerfile [hunk 4]
```diff
@@ -26,3 +26,19 @@ if github.pr_body.include?("* [ ]") && !github.pr_title.include?("[WIP]")
 end
 
 message "Pull Request is marked as Work in Progress" if github.pr_title.include? "[WIP]"
+
+junit.parse "output.xml"
+junit.report
```
Comment (ananyo2012) [inline Dangerfile]: Also we no longer need this line to show the failed test in a separate table as we are displaying it manually.

**Comment by jywarren** [comment]
Can you go ahead and try the fixes? Thanks!

**Comment by jywarren** [comment]
Ah, i found a moment and made the changes. Thanks for the catches!

**Comment by jywarren** [comment]
Still not quite right -- should be linking to the `danger-report` branch and not master. Ideas?

**Comment by jywarren** [comment]
used `branch_for_head` instead of `branch_for_merge`

**Comment by ananyo2012** [comment]
Great this is working now! We just have to remove the test now.

**Comment by jywarren** [comment]
But should we add a timestamp to the `output.xml` filename, and maybe call it `ci_reports.xml` -- and store it in the `/tests/reports/` folder? Or it will pile up messily?

**Comment by ananyo2012** [comment]
Adding time stamps maybe a good idea. But I think it isn't needed. Every time CI Reporter runs it regenerates all test results and the line `echo -e '<?xml version="1.0" encoding="UTF-8"?>' > output.xml`  ensures everything in `output.xml` is overwritten. The `>` symbol is used to overwrite content in a file while the `>>` symbol appends data to a file which is done in the next step.

**Comment by jywarren** [comment]
And i guess this is only used in the CI environment, so it won't create hanging `output.xml` files for regular developers. OK, then I'll merge in a sec.

Thanks, Ananyo! This one is really exciting. Shall we submit it to http://danger.systems/ruby/ as an example?

**Comment by ananyo2012** [comment]
Yes perhaps we can! Thanks to you for figuring out a way to display the errors :) This will really help beginners to make good PRs!

**Pull request merged**